core/src/main/java/org/elasticsearch/common/xcontent/ObjectParser.java

```diff
@@ -214,6 +222,89 @@ public void declareField(Parser<Value, Context> p, ParseField parseField, ValueT
         declareField((p, v, c) -> consumer.accept(v, parser.parse(p, c)), parseField, type);
     }
 
+    /**
+     * Declares a parser for fields where the exact field name is not known when
+     * creating the ObjectParser. In order for this field name to match, it has
+     * to be accepted by a provided predicate. As an example, in the aggregation
+     * output parsing for the high level java rest client we have things like:
+     *
+     * <pre>
+     *    "aggregations" : {
+     *      "terms#genres" : {  // aggregation type and arbitrary name
+     *         "doc_count_error_upper_bound": 0,
+     *         "sum_other_doc_count": 0,
+     *         "buckets" : [
+     *             {
+     *                 "key" : "jazz",
+     *                 "doc_count" : 10
+     *                 "sum#total_number_of_ratings": // aggregation type and arbitrary name
+     *                     "value": 2691
+     *                 }
+     *             },
+     *             [...]
+     *         ]
+     *      }
+     *    }
+     * </pre>
+     *
+     * Since field names are arbitrary in these cases, we cannot match them with
+     * a ParseField like we do in other places when using ObjectParser. This
+     * method can be used to register a special parser for a field name that
+     * matches the provided predicate.
+     *
+     * @param consumer
+     *            handle the values once they have been parsed
+     * @param parser
+     *            parses each nested object
+     * @param fieldNamePredicate
+     *            a predicate that returns true if the provided parser should
+     *            handle this field
+     * @param type
+     *            the accepted values for this field
+     */
+    public <T> void declareMatchFieldParser(BiConsumer<Value, T> consumer, ContextParser<Context, T> parser,
+            Predicate<String> fieldNamePredicate, ValueType type) {
+        Objects.requireNonNull(consumer, "[consumer] is required");
+        Objects.requireNonNull(parser, "[parser] is required");
+        Objects.requireNonNull(fieldNamePredicate, "[fieldNameMatcher] is required");
+        Objects.requireNonNull(type, "[type] is required");
+        this.matchField
+                .set(new MatchField(new MatchFieldParser((p, v, c) -> consumer.accept(v, parser.parse(p, c)), type), fieldNamePredicate));
+    }
+
+    private class MatchField {
+
+        private final FieldParser parser;
+        private final Predicate<String> predicate;
+
+        private MatchField(final FieldParser matchFieldParser, final Predicate<String> matchFieldPredicate) {
+            this.parser = matchFieldParser;
+            this.predicate = matchFieldPredicate;
+        }
+    }
+
+    private class MatchFieldParser extends FieldParser {
```

Comment:
I wonder why we need this class at this point. I guess it is because we don't provide a ParseField and the parseField.match part in assertSupports would not work. Would it be an idea to change parse field to always work with a predicate internally, and allow to pass the predicate in through a new constructor?